README.md
# 2020 — GPT-2 Generation + Defensive Distillation

Two threads of 2020 in one project: open-ended text generation with GPT-2, and
defensive distillation as a (now-discredited) defense against adversarial examples.

## What's here

- `src/gpt2_generate.py` — load GPT-2 medium, sample with top-k / top-p
- `src/defensive_distill.py` — train a teacher, then distill into a student at high
  softmax temperature; show how it appears to defend against FGSM but breaks under PGD
  with the right loss formulation

## Run

```bash
pip install -r requirements.txt
python src/gpt2_generate.py --prompt "The most important security risk of LLMs is"
python src/defensive_distill.py
```



CHANGELOG.md
# Changelog

## [0.3.0] — 2020-12-12
- Pin transformers <4.0 — 4.0 broke API

## [0.2.0] — 2020-10-25
- Why-it-fails docs (gradient masking explanation)
- Logit-difference probe demonstrating masking

## [0.1.0] — 2020-06-23
- Teacher/student distillation training
- GPT-2 generation with top-k/top-p sampling


tests/test_text_attacks.py
"""Tests for text-domain attacks. Uses a deterministic mock predictor so
we can verify attack mechanics without loading a transformer."""
import unittest

from src.text_attacks import (
    AttackResult,
    char_noise,
    importance_ranked_deletion,
    synonym_swap_attack,
    word_importance,
)


def make_keyword_predictor(target_words):
    """Returns a predict_proba(text) that flips on the presence of any
    target word. Class 0 = absent, class 1 = present.
    """
    target_set = set(target_words)
    def predict(text):
        words = set(text.lower().split())
        if words & target_set:
            return [0.05, 0.95]
        return [0.95, 0.05]
    return predict


class TestCharNoise(unittest.TestCase):
    def test_changes_text(self):
        out = char_noise("hello world this is fine", n_perturbations=5, seed=0)
        self.assertNotEqual(out, "hello world this is fine")

    def test_deterministic_with_seed(self):
        a = char_noise("hello world", n_perturbations=3, seed=42)
        b = char_noise("hello world", n_perturbations=3, seed=42)
        self.assertEqual(a, b)


class TestWordImportance(unittest.TestCase):
    def test_target_word_is_most_important(self):
        predict = make_keyword_predictor(['phishing'])
        scores = word_importance(predict, "this is a phishing example email", 1)
        # Removing "phishing" should drop class-1 prob the most
        words = "this is a phishing example email".split()
        idx = words.index('phishing')
        self.assertGreater(scores[idx], max(s for i, s in enumerate(scores) if i != idx))


class TestImportanceRankedDeletion(unittest.TestCase):
    def test_succeeds_when_target_word_present(self):
        predict = make_keyword_predictor(['phishing'])
        result = importance_ranked_deletion(
            predict, "alert: phishing detected here", 1, max_deletions=2
        )
        self.assertTrue(result.succeeded)
        self.assertNotIn('phishing', result.adversarial_text.lower())

    def test_fails_gracefully(self):
        predict = make_keyword_predictor(['phishing'])
        # Original prediction is class 0 (no target word); attack to flip
        # to class 1 by deleting words is impossible.
        result = importance_ranked_deletion(
            predict, "totally benign innocuous text", 0, max_deletions=3
        )
        self.assertFalse(result.succeeded)


class TestSynonymSwap(unittest.TestCase):
    def test_swaps_when_synonym_drops_score(self):
        predict = make_keyword_predictor(['great'])
        synonyms = {'great': ['mediocre']}  # "mediocre" not in target_set
        result = synonym_swap_attack(
            predict, "today was a great day", 1, synonyms, max_swaps=1
        )
        self.assertTrue(result.succeeded)
        self.assertIn('mediocre', result.adversarial_text)


if __name__ == '__main__':
    unittest.main()


src/text_attacks.py
"""Text-domain adversarial attacks against a fine-tuned classifier.

Adapts L-inf intuitions to discrete text. None of these are sound under
gradient masking — they're synonym/typo perturbations that work well
against classifiers without robust training.

Implements:
- TextFooler-style synonym swap (Jin et al. 2020)
- Random character noise (insertion / deletion / swap)
- Importance-ranked word deletion

The point is to compare a clean model vs. one trained with defensive
distillation: distillation reduces gradient magnitude but does NOT
actually defend against these decision-boundary-finding attacks.
"""
from __future__ import annotations

import random
import string
from dataclasses import dataclass
from typing import Callable, List, Optional


@dataclass
class AttackResult:
    original_text: str
    adversarial_text: str
    queries: int
    succeeded: bool
    original_pred: int
    final_pred: int


def char_noise(text: str, n_perturbations: int = 3, seed: int = 0) -> str:
    """Random character-level noise: insertion / deletion / swap."""
    rng = random.Random(seed)
    chars = list(text)
    for _ in range(n_perturbations):
        if not chars:
            break
        op = rng.choice(['insert', 'delete', 'swap'])
        i = rng.randrange(len(chars))
        if op == 'insert':
            chars.insert(i, rng.choice(string.ascii_lowercase))
        elif op == 'delete':
            del chars[i]
        elif op == 'swap' and i + 1 < len(chars):
            chars[i], chars[i+1] = chars[i+1], chars[i]
    return ''.join(chars)


def word_importance(
    predict_proba: Callable[[str], List[float]],
    text: str,
    target_class: int,
) -> List[float]:
    """For each word, score importance as drop in target-class prob when removed."""
    words = text.split()
    base = predict_proba(text)[target_class]
    scores = []
    for i in range(len(words)):
        masked = ' '.join(words[:i] + words[i+1:])
        new = predict_proba(masked)[target_class]
        scores.append(base - new)
    return scores


def importance_ranked_deletion(
    predict_proba: Callable[[str], List[float]],
    text: str,
    original_pred: int,
    max_deletions: int = 5,
) -> AttackResult:
    """Greedy attack: delete words in importance order until prediction flips."""
    words = text.split()
    queries = 0
    scores = word_importance(predict_proba, text, original_pred)
    queries += len(words) + 1
    order = sorted(range(len(words)), key=lambda i: -scores[i])

    current_text = text
    for k in range(1, max_deletions + 1):
        keep = sorted(order[k:])
        candidate = ' '.join(words[i] for i in keep)
        probs = predict_proba(candidate)
        queries += 1
        new_pred = max(range(len(probs)), key=lambda j: probs[j])
        if new_pred != original_pred:
            return AttackResult(
                original_text=text,
                adversarial_text=candidate,
                queries=queries,
                succeeded=True,
                original_pred=original_pred,
                final_pred=new_pred,
            )
    return AttackResult(
        original_text=text,
        adversarial_text=current_text,
        queries=queries,
        succeeded=False,
        original_pred=original_pred,
        final_pred=original_pred,
    )


def synonym_swap_attack(
    predict_proba: Callable[[str], List[float]],
    text: str,
    original_pred: int,
    synonym_dict: dict,
    max_swaps: int = 5,
) -> AttackResult:
    """TextFooler-style synonym replacement.

    `synonym_dict` maps each candidate word to a list of synonyms. In
    practice you'd use counter-fitted GloVe + a sense filter; here we
    parameterize so callers can plug in whatever vocabulary they want.
    """
    words = text.split()
    queries = 0
    base_scores = word_importance(predict_proba, text, original_pred)
    queries += len(words) + 1
    ranked = sorted(range(len(words)), key=lambda i: -base_scores[i])

    current = list(words)
    for idx in ranked[:max_swaps]:
        word = current[idx].lower().strip(string.punctuation)
        candidates = synonym_dict.get(word, [])
        best_drop = 0.0
        best_word = None
        for cand in candidates:
            trial = list(current)
            trial[idx] = cand
            probs = predict_proba(' '.join(trial))
            queries += 1
            drop = base_scores[idx] - (probs[original_pred] - 0.0)
            if drop > best_drop:
                best_drop = drop
                best_word = cand
        if best_word:
            current[idx] = best_word
            probs = predict_proba(' '.join(current))
            queries += 1
            new_pred = max(range(len(probs)), key=lambda j: probs[j])
            if new_pred != original_pred:
                return AttackResult(
                    original_text=text,
                    adversarial_text=' '.join(current),
                    queries=queries,
                    succeeded=True,
                    original_pred=original_pred,
                    final_pred=new_pred,
                )
    return AttackResult(
        original_text=text,
        adversarial_text=' '.join(current),
        queries=queries,
        succeeded=False,
        original_pred=original_pred,
        final_pred=original_pred,
    )


src/gpt2_generate.py
"""GPT-2 generation with top-k / top-p sampling. transformers 3.x."""
import argparse

import torch
from transformers import GPT2LMHeadModel, GPT2Tokenizer


def generate(prompt, model_name='gpt2-medium', max_length=120,
             top_k=50, top_p=0.92, temperature=0.9, n=3):
    device = 'cuda' if torch.cuda.is_available() else 'cpu'
    tok = GPT2Tokenizer.from_pretrained(model_name)
    model = GPT2LMHeadModel.from_pretrained(model_name).to(device)
    model.eval()

    ids = tok.encode(prompt, return_tensors='pt').to(device)
    out = model.generate(
        ids,
        max_length=max_length,
        do_sample=True,
        top_k=top_k,
        top_p=top_p,
        temperature=temperature,
        num_return_sequences=n,
        pad_token_id=tok.eos_token_id,
    )
    for i, seq in enumerate(out):
        print('--- sample %d ---' % (i + 1))
        print(tok.decode(seq, skip_special_tokens=True))


if __name__ == '__main__':
    p = argparse.ArgumentParser()
    p.add_argument('--prompt', default='The most important security risk of LLMs is')
    p.add_argument('--model', default='gpt2-medium')
    args = p.parse_args()
    generate(args.prompt, model_name=args.model)


src/defensive_distill.py
"""Defensive distillation demo on MNIST.

Step 1: Train teacher with high-temperature softmax.
Step 2: Distill student from teacher's soft labels.
Step 3: FGSM appears to fail (gradients vanish under high T).
Step 4: Carlini-Wagner-style attack with logit-targeted loss bypasses it.

Lesson: gradient masking != robustness. This was understood by 2020 but the
worked example is still the cleanest way to grok why.
"""
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import DataLoader
from torchvision import datasets, transforms

T = 20.0  # distillation temperature


class CNN(nn.Module):
    def __init__(self):
        super().__init__()
        self.c1 = nn.Conv2d(1, 32, 3, padding=1)
        self.c2 = nn.Conv2d(32, 64, 3, padding=1)
        self.fc1 = nn.Linear(64 * 7 * 7, 128)
        self.fc2 = nn.Linear(128, 10)

    def forward(self, x):
        x = F.max_pool2d(F.relu(self.c1(x)), 2)
        x = F.max_pool2d(F.relu(self.c2(x)), 2)
        return self.fc2(F.relu(self.fc1(x.flatten(1))))


def loaders(bs=128):
    tfm = transforms.ToTensor()
    tr = datasets.MNIST('./data', train=True, download=True, transform=tfm)
    te = datasets.MNIST('./data', train=False, download=True, transform=tfm)
    return DataLoader(tr, batch_size=bs, shuffle=True), DataLoader(te, batch_size=bs)


def train(model, loader, epochs=2, T_=1.0, soft_targets=None, dev='cpu'):
    opt = torch.optim.Adam(model.parameters(), lr=1e-3)
    for ep in range(epochs):
        model.train()
        for i, (x, y) in enumerate(loader):
            x, y = x.to(dev), y.to(dev)
            logits = model(x) / T_
            if soft_targets is None:
                loss = F.cross_entropy(logits, y)
            else:
                # distillation loss: KL between student and teacher soft logits
                target = soft_targets(x) / T_
                loss = F.kl_div(F.log_softmax(logits, -1),
                                F.softmax(target, -1), reduction='batchmean')
            opt.zero_grad()
            loss.backward()
            opt.step()


def fgsm(model, x, y, eps, T_=1.0):
    x = x.clone().detach().requires_grad_(True)
    loss = F.cross_entropy(model(x) / T_, y)
    g = torch.autograd.grad(loss, x)[0]
    return (x + eps * g.sign()).clamp(0, 1).detach()


def cw_logit_attack(model, x, y, eps=0.3, alpha=0.01, steps=40):
    """Direct logit-difference attack — bypasses high-T gradient masking."""
    x_orig = x.clone().detach()
    x = x_orig + torch.empty_like(x).uniform_(-eps, eps)
    x = x.clamp(0, 1)
    for _ in range(steps):
        x = x.detach().requires_grad_(True)
        logits = model(x)  # no /T scaling — operate on raw logits
        true_logit = logits.gather(1, y.unsqueeze(1)).squeeze(1)
        other_max = logits.scatter(1, y.unsqueeze(1), -1e9).max(1)[0]
        loss = (true_logit - other_max).mean()  # minimize -> attack
        g = torch.autograd.grad(loss, x)[0]
        x = (x - alpha * g.sign())
        x = (x_orig + (x - x_orig).clamp(-eps, eps)).clamp(0, 1)
    return x.detach()


def evaluate(model, attack_fn, loader, dev):
    model.eval()
    correct = total = 0
    for x, y in loader:
        x, y = x.to(dev), y.to(dev)
        x_adv = attack_fn(model, x, y)
        with torch.no_grad():
            correct += (model(x_adv).argmax(1) == y).sum().item()
        total += y.size(0)
    return correct / total


def main():
    dev = 'cuda' if torch.cuda.is_available() else 'cpu'
    train_loader, test_loader = loaders()

    print('training teacher (T=%.0f)...' % T)
    teacher = CNN().to(dev)
    train(teacher, train_loader, epochs=2, T_=T, dev=dev)

    print('distilling student...')
    student = CNN().to(dev)
    train(student, train_loader, epochs=2, T_=T, dev=dev,
          soft_targets=lambda x: teacher(x).detach())

    print('FGSM @ eps=0.3 (T-scaled): acc =',
          evaluate(student, lambda m, x, y: fgsm(m, x, y, 0.3, T_=T),
                   test_loader, dev))
    print('CW-style logit attack:    acc =',
          evaluate(student, lambda m, x, y: cw_logit_attack(m, x, y),
                   test_loader, dev))


if __name__ == '__main__':
    main()


src/eval_suite.py
"""Evaluation harness comparing clean vs. defensively-distilled classifiers.

The setup uses a simple sentiment / toxicity classifier (loaded from a
HuggingFace checkpoint) as the "victim". We then run text attacks against
both the victim and a distilled student, and report:

  - clean accuracy (sanity check distillation didn't tank utility)
  - attack success rate per attack type
  - mean queries to flip
  - the gradient-masking diagnostic: ratio of FGSM-style probability
    distance vs. importance-ranked deletion success — these should
    track each other; if FGSM looks defended but ranked-deletion still
    flips, you have masking, not robustness.
"""
import argparse
import json
import os
from dataclasses import asdict
from statistics import mean

import torch
from transformers import AutoTokenizer, AutoModelForSequenceClassification

from src.text_attacks import (
    AttackResult,
    char_noise,
    importance_ranked_deletion,
    synonym_swap_attack,
)


SIMPLE_SYNONYMS = {
    'good': ['fine', 'decent', 'okay', 'pleasant'],
    'bad': ['poor', 'awful', 'terrible', 'unfortunate'],
    'great': ['excellent', 'fantastic', 'wonderful'],
    'terrible': ['horrible', 'dreadful', 'atrocious'],
    'love': ['adore', 'enjoy', 'appreciate'],
    'hate': ['despise', 'loathe', 'detest'],
    'best': ['finest', 'top', 'greatest'],
    'worst': ['poorest', 'lousiest'],
}


def make_predict(model, tokenizer, device, max_len=128):
    def predict_proba(text):
        enc = tokenizer(
            text,
            return_tensors='pt',
            truncation=True,
            padding='max_length',
            max_length=max_len,
        ).to(device)
        with torch.no_grad():
            out = model(**enc)
        probs = torch.softmax(out.logits, dim=-1)[0].cpu().tolist()
        return probs
    return predict_proba


def evaluate(model_name: str, examples: list, max_attacks: int = 50) -> dict:
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    model = AutoModelForSequenceClassification.from_pretrained(model_name).to(device)
    model.eval()
    predict = make_predict(model, tokenizer, device)

    n = 0
    n_correct = 0
    delete_results = []
    synonym_results = []
    char_flips = 0
    for text, label in examples[:max_attacks]:
        probs = predict(text)
        pred = max(range(len(probs)), key=lambda i: probs[i])
        n += 1
        if pred == label:
            n_correct += 1
        else:
            continue  # only attack examples the model already gets right

        delete_results.append(
            importance_ranked_deletion(predict, text, pred)
        )
        synonym_results.append(
            synonym_swap_attack(predict, text, pred, SIMPLE_SYNONYMS)
        )
        noisy = char_noise(text, n_perturbations=5, seed=hash(text) & 0xFFFF)
        np_probs = predict(noisy)
        if max(range(len(np_probs)), key=lambda i: np_probs[i]) != pred:
            char_flips += 1

    return {
        'n_examples': n,
        'clean_accuracy': n_correct / n if n else float('nan'),
        'deletion_success_rate':
            sum(r.succeeded for r in delete_results) / max(len(delete_results), 1),
        'deletion_mean_queries':
            mean([r.queries for r in delete_results]) if delete_results else 0,
        'synonym_success_rate':
            sum(r.succeeded for r in synonym_results) / max(len(synonym_results), 1),
        'char_noise_flip_rate':
            char_flips / max(n_correct, 1),
        'sample_results': {
            'deletion': [asdict(r) for r in delete_results[:3]],
            'synonym': [asdict(r) for r in synonym_results[:3]],
        },
    }


def main():
    p = argparse.ArgumentParser()
    p.add_argument('--clean-model', required=True,
                   help='HF checkpoint of the clean classifier')
    p.add_argument('--distilled-model', required=True,
                   help='HF checkpoint of the distilled student')
    p.add_argument('--examples-json', required=True,
                   help='JSON list of [text, label] pairs')
    p.add_argument('--out', default='results/eval_suite.json')
    p.add_argument('--max-attacks', type=int, default=50)
    args = p.parse_args()

    with open(args.examples_json) as f:
        examples = json.load(f)

    clean = evaluate(args.clean_model, examples, args.max_attacks)
    distilled = evaluate(args.distilled_model, examples, args.max_attacks)

    summary = {
        'clean': clean,
        'distilled': distilled,
        'gradient_masking_diagnostic':
            ('Char-noise flips dropped' if distilled['char_noise_flip_rate'] <
                clean['char_noise_flip_rate']
             else 'Char-noise unchanged') +
            ' BUT ' +
            ('deletion success similar' if abs(distilled['deletion_success_rate'] -
                clean['deletion_success_rate']) < 0.10
             else 'deletion success dropped'),
    }
    os.makedirs(os.path.dirname(args.out), exist_ok=True)
    with open(args.out, 'w') as f:
        json.dump(summary, f, indent=2)
    print(json.dumps(summary, indent=2))


if __name__ == '__main__':
    main()
